File: source/server/schedule/schedule.go
package schedule

import (
	"fmt"
	"time"

	"cloud.google.com/go/civil"
)

type Rule struct {
	Type  string `json:"type"`
	Every int    `json:"every,omitempty"`
	Unit  string `json:"unit,omitempty"`
	Day   int    `json:"day,omitempty"`
	Month int    `json:"month,omitempty"`
}

func NextDueDate(rule Rule, baseline civil.Date) (civil.Date, error) {
	if !baseline.IsValid() {
		return civil.Date{}, fmt.Errorf("invalid baseline date")
	}

	switch rule.Type {
	case "relative":
		if rule.Every <= 0 {
			return civil.Date{}, fmt.Errorf("every must be a positive integer")
		}

		switch rule.Unit {
		case "days":
			return baseline.AddDays(rule.Every), nil
		case "weeks":
			return baseline.AddDays(rule.Every * 7), nil
		case "months":
			return addMonths(baseline, rule.Every)
		case "years":
			return clampDate(baseline.Year+rule.Every, baseline.Month, baseline.Day)
		default:
			return civil.Date{}, fmt.Errorf("unsupported relative unit %q", rule.Unit)
		}
	case "fixed-monthly":
		current, err := clampDate(baseline.Year, baseline.Month, rule.Day)
		if err != nil {
			return civil.Date{}, err
		}
		if current.After(baseline) {
			return current, nil
		}
		nextMonth := baseline.Month + 1
		nextYear := baseline.Year
		if nextMonth > time.December {
			nextMonth = time.January
			nextYear++
		}
		return clampDate(nextYear, nextMonth, rule.Day)
	case "fixed-yearly":
		if rule.Month < 1 || rule.Month > 12 {
			return civil.Date{}, fmt.Errorf("invalid month")
		}
		month := time.Month(rule.Month)
		if rule.Day > daysInMonth(2000, month) {
			return civil.Date{}, fmt.Errorf("invalid day for month")
		}
		current, err := clampDate(baseline.Year, month, rule.Day)
		if err != nil {
			return civil.Date{}, err
		}
		if current.After(baseline) {
			return current, nil
		}
		return clampDate(baseline.Year+1, month, rule.Day)
	default:
		return civil.Date{}, fmt.Errorf("unsupported schedule type %q", rule.Type)
	}
}

File: source/packages/core/src/schedule.ts
import { addCalendarDays, clampLocalDate, daysInMonth, parseLocalDate } from "./date";
import type { LocalDate, ScheduleRule } from "./types";

function assertPositiveInteger(value: number, field: string): void {
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`${field} must be a positive integer`);
  }
}

function nextFixedMonthly(day: number, baseline: LocalDate): LocalDate {
  assertPositiveInteger(day, "day");

  if (day > 31) {
    throw new Error("day must not be greater than 31");
  }

  const parts = parseLocalDate(baseline);
  const currentMonthOccurrence = clampLocalDate(parts.year, parts.month, day);

  if (currentMonthOccurrence > baseline) {
    return currentMonthOccurrence;
  }

  const nextMonth = parts.month === 12 ? 1 : parts.month + 1;
  const nextYear = parts.month === 12 ? parts.year + 1 : parts.year;

  return clampLocalDate(nextYear, nextMonth, day);
}

function nextFixedYearly(month: number, day: number, baseline: LocalDate): LocalDate {
  assertPositiveInteger(month, "month");
  assertPositiveInteger(day, "day");

  if (month > 12) {
    throw new Error("month must not be greater than 12");
  }

  // Validate against a leap year so fixed-yearly rules may target 2/29.
  // clampLocalDate clamps it to 2/28 in non-leap years.
  if (day > 31 || day > daysInMonth(2000, month)) {
    throw new Error("day is invalid for month");
  }

  const parts = parseLocalDate(baseline);
  const currentYearOccurrence = clampLocalDate(parts.year, month, day);

  if (currentYearOccurrence > baseline) {
    return currentYearOccurrence;
  }

  return clampLocalDate(parts.year + 1, month, day);
}

function addRelativeMonths(baseline: LocalDate, monthsToAdd: number): LocalDate {
  const parts = parseLocalDate(baseline);
  const zeroBasedTargetMonth = parts.month - 1 + monthsToAdd;
  const year = parts.year + Math.floor(zeroBasedTargetMonth / 12);
  const month = (zeroBasedTargetMonth % 12) + 1;

  return clampLocalDate(year, month, parts.day);
}

export function nextDueDate(rule: ScheduleRule, baseline: LocalDate): LocalDate {
  parseLocalDate(baseline);

  if (rule.type === "fixed-monthly") {
    return nextFixedMonthly(rule.day, baseline);
  }

  if (rule.type === "fixed-yearly") {
    return nextFixedYearly(rule.month, rule.day, baseline);
  }

  assertPositiveInteger(rule.every, "every");

  switch (rule.unit) {
    case "days":
      return addCalendarDays(baseline, rule.every);
    case "weeks":
      return addCalendarDays(baseline, rule.every * 7);
    case "months":
      return addRelativeMonths(baseline, rule.every);
    case "years": {
      const parts = parseLocalDate(baseline);
      return clampLocalDate(parts.year + rule.every, parts.month, parts.day);
    }
  }
}
